tests/integration/test_charm_ipa.py

```diff
@@ -0,0 +1,172 @@
+    _, info, _ = await ops_test.juju("show-unit", unit)
+
+    remote_unit = REQUIRER_MOCK_APP_NAME + "/0"
+    requirer_unit_databag = yaml.safe_load(info)[unit]["relation-info"][0]["application-data"][
+        "data"
+    ]
+    model_name = ops_test.model_name
+    expected_requirer_unit_data = {
+        "host": "0.0.0.0",
+        "model": model_name,
+        "name": remote_unit,
+        "port": PORT,
+    }
+
+    # let's ensure it matches our own schema
+    _validate_data(expected_requirer_unit_data, INGRESS_REQUIRES_UNIT_SCHEMA)
+
+    ingress_data = yaml.safe_load(requirer_unit_databag)
+    assert ingress_data == expected_requirer_unit_data
+
+
+async def test_provider_app_databag(ops_test: OpsTest):
+    remote_unit = REQUIRER_MOCK_APP_NAME + "/0"
+    _, info, _ = await ops_test.juju("show-unit", remote_unit)
+    model_name = ops_test.model_name
+
+    expected_requirer_app_data = {
+        "ingress": {"url": f"http://{HOSTNAME}:{PORT}/{model_name}-{REQUIRER_MOCK_APP_NAME}"}
+    }
+
+    assert_app_databag_equals(
+        info, remote_unit, expected_requirer_app_data, INGRESS_PROVIDES_APP_SCHEMA
+    )
+
+
+async def test_scale_up_requirer(ops_test: OpsTest):
+    # add two units of requirer mock
+    await ops_test.juju("add-unit", REQUIRER_MOCK_APP_NAME, "-n2")
+    await assert_status_reached(
+        ops_test,
+        apps=[REQUIRER_MOCK_APP_NAME],
+        status="active",
+        raise_on_blocked=False,
+        wait_for_exact_units=3,
+    )
+
+
+async def test_traefik_relation_data_after_upscale(ops_test: OpsTest):
+    remote_unit = REQUIRER_MOCK_APP_NAME + "/0"
+    _, info, _ = await ops_test.juju("show-unit", remote_unit)
+    model_name = ops_test.model_name
+
+    expected_requirer_app_data = {
+        "ingress": {"url": f"http://{HOSTNAME}:{PORT}/{model_name}-{REQUIRER_MOCK_APP_NAME}"}
+    }
+
+    assert_app_databag_equals(
+        info, remote_unit, expected_requirer_app_data, INGRESS_PROVIDES_APP_SCHEMA
+    )
+
+
+async def test_scale_down_requirer(ops_test: OpsTest):
+    # remove one unit; there should be two left
+    await ops_test.juju("remove-unit", REQUIRER_MOCK_APP_NAME, "--num-units", "1")
+    await assert_status_reached(
+        ops_test, apps=[REQUIRER_MOCK_APP_NAME], status="active", wait_for_exact_units=2
+    )
+
+
+async def test_traefik_relation_data_after_downscale(ops_test: OpsTest):
+    remote_unit = REQUIRER_MOCK_APP_NAME + "/0"
+    _, info, _ = await ops_test.juju("show-unit", remote_unit)
+    model_name = ops_test.model_name
+
+    expected_requirer_app_data = {
+        "ingress": {"url": f"http://{HOSTNAME}:{PORT}/{model_name}-{REQUIRER_MOCK_APP_NAME}"}
+    }
+
+    assert_app_databag_equals(
+        info, remote_unit, expected_requirer_app_data, INGRESS_PROVIDES_APP_SCHEMA
+    )
+
+
+# cleanup before closing this test module: unrelate applications, scale requirer
+# mock back down to 1 and check final status
+async def test_reset_to_initial_state(ops_test):
```

Comment:
Could do with a test that does relate, unrelate then relate again and check things are consistent?

Reply:
done!